packages/dcos-sdk/src/NavigationService.ts

```diff
@@ -0,0 +1,80 @@
+import { injectable, inject, named } from "inversify";
+import { EventEmitter } from "events";
+import { IExtensionProvider, ExtensionProvider } from "plugin-baby";
+
+const NAVIGATION_CHANGE = Symbol("NAVIGATION_CHANGE");
+
+interface INavigationElement {
+  children?: INavigationElement[],
+  parent?: string,
+  path?: string
+}
+
+export const NavigationServiceExtension = Symbol("NavigationServiceExtension");
+
+export interface INavigationExtension {
+  getElements(): INavigationElement[];
+}
+
+function unflatten(array: INavigationElement[], parent: INavigationElement = {}) {
+  const children = array.filter((child: INavigationElement) =>
+    child.parent === parent.path
+  );
+
+  if (children.length !== 0) {
+    if (parent.path) {
+      parent.children = children;
+    }
+
+    children.forEach((child: INavigationElement) => unflatten(array, child));
+  }
+
+  return children;
+}
+
+@injectable()
+export default class NavigationService {
+  private extensionProvider: IExtensionProvider<INavigationExtension>;
+  private eventEmmiter: EventEmitter;
+  private definition: INavigationElement[];
+
+  constructor(
+    @inject(ExtensionProvider)
+    @named(NavigationServiceExtension)
+    extensionProvider: IExtensionProvider<INavigationExtension>
+  ) {
+    this.extensionProvider = extensionProvider;
+
+    this.eventEmmiter = new EventEmitter();
+    this.definition = [];
+
+    this.extensionProvider.subscribe(() => {
+      this.definition = [];
+      this.eventEmmiter.emit(NAVIGATION_CHANGE);
+    });
+  }
+
+  public on(type: symbol, callback: () => void) {
```

Comment:
I saw at some places that we just extend from EventEmitter, why did you choose to do it this way instead? Or was the other occasion legacy code?

Reply:
In my previous attempts it didn't work nicely. Extending `EventEmitter` would mean annotation it as `injectable` and I discovered some nasty side effects when I tried. So I fell back to just wrapping it.

Reply:
Okay, makes sense 👍